lib/gssapi/mech/gss_cred.c

```diff
@@ -34,12 +34,70 @@
 #include "mech_locl.h"
 #include <krb5.h>
 
+static OM_uint32
+export_oid_set(OM_uint32 *minor_status,
+	       gss_const_OID mech,
+	       gss_const_OID_set oids,
+	       gss_buffer_t token)
+{
+    krb5_error_code ret;
+    krb5_storage *sp;
+    krb5_data data;
+    size_t i, len;
+
+    _mg_buffer_zero(token);
+
+    sp = krb5_storage_emem();
+    if (sp == NULL) {
+	ret = ENOMEM;
+	goto out;
+    }
+
+    data.length = mech->length;
+    data.data = mech->elements;
+
+    ret = krb5_store_data(sp, data);
+    if (ret)
+	goto out;
+
+    for (i = 0, len = 0; i < oids->count; i++)
+	len += 4 + oids->elements[i].length;
```

Comment:
Why not just store the count of OIDs?  Why do we need to add all these other things?

Reply:
Because there is (frustratingly) an asymmetry between `gss_import_cred()` and gss_export_cred()`. As noted in the source, `gss_export_cred()` expects the mechanism's `gm_export_cred()` (which we are emulating with `export_oid_set()`) to return:

```
mech-len mech-data cred-len cred-data
```

to be written to directly to `sp`; `gss_import_cred()` however, peeks inside this and only passes `mech-data` to `gm_import_cred()`. A similar asymmetry is present when importing and exporting security contexts. So, we're just playing by the rules I'm afraid.